### PR TITLE
fix(runner): use ConPTY for reliable Windows reloads

### LINKED ISSUE #228: Windows: Uvicorn reload stalls after the first source change
## Description

On Windows 10, a Python backend launched through Gomander with Uvicorn reload mode reloads successfully after the first source change, but subsequent changes no longer complete the reload cycle.

## Steps to reproduce

1. Configure a command with:

       uv run dev --port 8000 --reload --timeout-graceful-shutdown 30

2. Start the command on Windows 10.
3. Modify a watched Python source file; the first reload succeeds.
4. Modify a watched source file again.

## Expected behavior

Every source change should stop the current worker and start a new one.

## Actual behavior

The first reload succeeds, but the second reload stalls and the backend is not restarted.

## Investigation

The issue was reproduced with Uvicorn 0.50.0 and watchfiles 1.2.0. Windows commands were launched through a hidden cmd.exe process with redirected stdout/stderr pipes. This does not preserve the console control semantics used by Uvicorn during repeated worker shutdown and restart cycles.

Running the command through Windows ConPTY preserves terminal behavior; three consecutive reloads complete successfully in the reproduction test.

## Proposed fix

Use ConPTY for Windows command execution, retain a fallback for systems where ConPTY is unavailable, and make process wait/shutdown ownership race-safe.

Fix PR: #227

### PR DESCRIPTION
**Pull request by ailtariel-lin**
## Summary

Fix repeated Python/Uvicorn reloads stalling on Windows 10 by running commands inside ConPTY. The previous hidden `cmd.exe` process with redirected pipes did not preserve the console control semantics Uvicorn relies on after the first reload.

## Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Test improvement
- [ ] Documentation update

## Changes Made

- Add a cross-platform command process abstraction with a Windows ConPTY implementation.
- Preserve ANSI output and command exit errors while filtering ConPTY host-only initialization sequences.
- Keep the existing hidden-process implementation as a fallback when ConPTY is unavailable.
- Make process waiting and shutdown single-owner and race-safe.
- Add Windows tests for terminal output semantics and non-zero command failures.
- Clean up infinite commands in the existing runner test.

## Testing

- [x] Reproduced the original failure with Uvicorn 0.50.0 and watchfiles 1.2.0 on Windows.
- [x] Verified three consecutive reloads through `DefaultRunner` after the fix.
- [x] `go test ./... -count=1`
- [x] `go test ./internal/runner -count=1`
- [x] `go vet ./internal/runner`
- [x] Linux runner cross-compilation
- [x] `pnpm run build`

`go test -race` was not available in the local Windows environment because CGO requires a C compiler, which is not installed.

## Checklist

- [x] My code follows the existing style of this project.
- [x] I have performed a self-review of my code.
- [x] I have added tests that prove the fix is effective.
- [x] New and existing unit tests pass locally with my changes.
- [x] My changes generate no new compiler or vet warnings.

## Related Issues

Closes #228

## Summary by CodeRabbit

- **New Features**
  - Improved command execution across Unix and Windows environments.
  - Added enhanced Windows terminal support for cleaner output and terminal-aware behavior.
  - Added support for configured terminal dimensions, environment paths, and forced color output.

- **Bug Fixes**
  - Improved handling of command failures and output cleanup.
  - Removed unwanted terminal control sequences from Windows command output.
  - Made process shutdown more reliable, escalating from graceful termination to forced termination when necessary.
  - Filtered known irrelevant process output lines.

### CODE REVIEW

File: internal/runner/proc_unix.go [hunk 1]
```diff
@@ -5,70 +5,41 @@ package runner
...
 	select {
-	case <-time.After(5 * time.Second):
-		// Force kill if graceful shutdown takes too long
-		return syscall.Kill(-cmd.Process.Pid, syscall.SIGKILL)
+	case <-timer.C:
+		return syscall.Kill(-process.PID(), syscall.SIGKILL)
 	case <-done:
 		return nil
 	}
 }
```
[the remark this answers is not in the document]
Comment (ailtariel-lin) [reply internal/runner/proc_unix.go:56]: Fixed in 82c2fbd. stopProcessGracefully now treats syscall.ESRCH from the initial SIGTERM and the fallback or timer SIGKILL as successful completion, while preserving every other signal error. Added Unix coverage for ESRCH and unrelated errors.

File: internal/runner/proc_windows.go [hunk 2]
```diff
@@ -3,81 +3,201 @@
...
+	// output pipe to EOF, so buffered tail output is not discarded.
+	windows.ClosePseudoConsole(windows.Handle(p.conPTY.Fd()))
+	_ = p.conPTY.InputPipe().Close()
+	// go-pty v0.2.2 exposes the exit code in ProcessState but does not turn a
+	// non-zero Windows exit into an error as exec.Cmd does.
+	if waitErr == nil && p.conPTYCommand.ProcessState != nil && !p.conPTYCommand.ProcessState.Success() {
+		return fmt.Errorf("exit status %d", p.conPTYCommand.ProcessState.ExitCode())
 	}
+	return waitErr
+}
```
[the remark this answers is not in the document]
Comment (ailtariel-lin) [reply internal/runner/proc_windows.go:166]: Verified against go-pty v0.2.2 and kept the current close order. ConPty.Close closes outPipe immediately, so calling it from Wait can discard buffered tail output. After the child exits, this implementation closes HPCON so conhost closes its writer, closes the input pipe, drains output to EOF, and the runner then closes the output reader through conPTYOutput.Close. All three resources are released without closing the read side before it drains.

File: internal/runner/runner.go [hunk 3]
```diff
@@ -161,18 +139,14 @@ func (c *DefaultRunner) RunCommand(command *domain.Command, environmentPaths []s
 			c.eventEmitter.EmitEvent(event.ProcessFinished, command.Id)
 		}()
 
-		// Wait for all pipes to finish
+		err := process.Wait()
 		scanWg.Wait()
```
[the remark this answers is not in the document]
Comment (ailtariel-lin) [reply internal/runner/runner.go:135]: Fixed in 82c2fbd. execProcess now creates caller-owned os.Pipe pairs, attaches only the write ends to the child, and closes the parent write copies after Start. Cmd.Wait therefore no longer owns or closes the reader sides, and readers can drain trailing output after process exit. Added a regression test that failed with file already closed before this change and now verifies the full tail payload after Wait.

**Comment by ailtariel-lin** [comment]
Addressed the remaining outside-diff and nitpick findings in 82c2fbd:

- Removed the published WaitGroup entirely. RunningCommand now exposes its pre-created done channel, which closes only after process waiting, output draining, cleanup, and the finished event complete.
- Added /bin/sh and cmd.exe defaults when SHELL or COMSPEC is empty.
- Closed the reader in TestWindowsProcessUsesTerminalOutput after ReadAll.
- Kept the initial ConPTY read line-oriented. DefaultRunner immediately feeds it to bufio.Scanner, so partial lines are intentionally not emitted; prefix-level buffering would not reduce user-visible latency in this pipeline.

Validation:
- go test ./internal/runner -count=1
- go test ./... -count=1
- go vet ./internal/runner
- Linux runner test cross-compilation
- Real DefaultRunner plus Uvicorn 0.50.0 and WatchFiles: three consecutive source changes each completed shutdown, worker replacement, and startup.

The PR is now linked to #228 with Closes #228.

**Comment by ailtariel-lin** [comment]
@coderabbitai review

**Comment by moisesrj97** [comment]
Thanks for this — the issue write-up and the Windows 10 repro are excellent, and that's work we can't do on demand.

We'd like to hold off on moving Windows execution to a PTY for now. It'd become the default for every command and changes a lot at once (live stdin nothing writes to, VT repaints wiping logs mid-run, go-pty + u-root on all platforms), and we're not yet sure it's needed — there's a suspect already in your diff. `proc_windows.go` used to set `CREATE_NEW_PROCESS_GROUP`, which disables Ctrl+C for the whole descendant tree. Nothing here ever sends a console control event, so it buys us nothing — and your branch deletes it. Some of the fix may be that one line.

Could you try, on your Windows box: (1) just dropping that flag on `main`; (2) re-running the manual repro on `main` with and without a browser connected (`--timeout-graceful-shutdown 30` plus a keep-alive can produce exactly this "first reload fine, later ones stall" shape)?

Separately, the process-lifecycle fix here is real and independent — on `main`, `StopProcessGracefully` and the wait goroutine can both call `cmd.Wait()`. We'd merge that as its own small PR with your tests, whatever happens with the rest.

**Comment by ailtariel-lin** [comment]
Following up with the requested validation and a narrower proposal.

## 1. Validation results

Test environment:

- Windows 10 Pro 22H2, build 19045, x64
- Python 3.11.5
- Uvicorn 0.50.0
- watchfiles 1.2.0
- `--timeout-graceful-shutdown 30`
- Three consecutive source changes per successful run

Results:

| Runner variant | Keep-alive client | Result |
| --- | --- | --- |
| `main` | No | WatchFiles detected the first change, Uvicorn printed `Reloading...`, and the first reload then stalled until the 40-second test timeout |
| `main` | Yes | Same first-reload stall |
| `main` with only `CREATE_NEW_PROCESS_GROUP` removed | No | Same first-reload stall |
| `main` using a hidden `CREATE_NEW_CONSOLE` instead | No | Same first-reload stall |
| ConPTY runner | No | Three reloads completed in about 3.2 seconds |
| ConPTY runner | Yes | Three reloads completed in about 3.4 seconds |

The keep-alive case used a persistent HTTP/1.1 connection during each file change. I also checked the failed runs for leftover process trees; the baseline variants did leave Uvicorn descendants behind, whereas the ConPTY runs completed and cleaned up normally.

So the original flag was a valid suspect, but removing it alone does not fix this reproduction.

## 2. Why the simple, dependency-free alternatives did not solve it

The relevant Uvicorn reload path on Windows calls:

```python
os.kill(self.process.pid, signal.CTRL_C_EVENT)
self.process.join()
```

On Windows, CPython maps this to `GenerateConsoleCtrlEvent(CTRL_C_EVENT, pid)`. The second argument is interpreted by Windows as a **process-group ID**, not an arbitrary process ID.

There are two overlapping Windows constraints here:

1. `CREATE_NEW_PROCESS_GROUP` disables Ctrl+C handling for the new group, so removing it was necessary to test.
2. More importantly, Microsoft documents that `CTRL_C_EVENT` cannot be limited to a specific process group. Supplying a non-zero group ID may return success while no process receives the event. See [GenerateConsoleCtrlEvent](https://learn.microsoft.com/en-us/windows/console/generateconsolectrlevent).

That explains why simply removing the creation flag did not change the observed result: it removes one blocker, but it does not repair the `CTRL_C_EVENT` targeting semantics used inside Uvicorn's reload supervisor.

I also tried creating a new console instead of a new process group. That still stalled. Redirected stdin/stdout pipes, creation flags, and console allocation do not give the child the same terminal path that ConPTY provides.

I do not mean that a dependency-free implementation is theoretically impossible. We could call the ConPTY APIs directly and implement the pseudoconsole handle, pipe, resize, startup-attribute-list, wait, and cleanup lifecycle ourselves. However, that would reproduce a substantial part of a PTY library inside Gomander, with more Windows-specific code and more lifecycle edge cases than using the existing dependency. Other alternatives would require changing Uvicorn, injecting a command-specific wrapper, or intercepting behavior inside the child process; none is a general runner fix.

Based on the failed minimal experiments, I could not find a **small and general** dependency-free change that restores the required terminal semantics. ConPTY is the only approach tested here that changes the observed behavior.

## 3. Revised, scoped approach

I reworked the proposal so ConPTY is no longer selected for every Windows environment.

Source references for the allowlist described above:

- The application-level allowlist is defined in [`cmd/gomander/application_config.go`](https://github.com/ailtariel-lin/gomander-app/blob/457b6817f85e60c0d503ec4856b985bdeb1cb419/cmd/gomander/application_config.go#L10-L16). It currently contains exactly one entry: `runner.HostEnvironmentWindows10`.
- The `HostEnvironment` identifier, `runner.Config.ConPTYEnvironments`, and membership check are defined in [`internal/runner/config.go`](https://github.com/ailtariel-lin/gomander-app/blob/457b6817f85e60c0d503ec4856b985bdeb1cb419/internal/runner/config.go#L3-L22).
- The Windows process selector consumes that configuration in [`internal/runner/proc_windows.go`](https://github.com/ailtariel-lin/gomander-app/blob/457b6817f85e60c0d503ec4856b985bdeb1cb419/internal/runner/proc_windows.go#L80-L99). The Windows 10 workstation classification and final allowlist/API gate are in [the same file](https://github.com/ailtariel-lin/gomander-app/blob/457b6817f85e60c0d503ec4856b985bdeb1cb419/internal/runner/proc_windows.go#L157-L176).

Therefore, expanding ConPTY to another host environment requires an explicit application-configuration change; detecting a different Windows version by itself does not enable it.

Application-level compatibility parameters are kept together and injected into the runner. The initial ConPTY allowlist contains only `windows10`. At runtime, ConPTY is selected only when all of these conditions hold:

- the detected host is a Windows workstation;
- the version is Windows 10 build 17763 through 21999;
- `windows10` is present in the configured allowlist; and
- `CreatePseudoConsole` is available.

Windows 11, Windows Server, Windows 10 before 1809, an unavailable ConPTY API, and any environment not in the allowlist retain the existing hidden `cmd /C` pipe path. ConPTY-only output filtering is also confined to the ConPTY implementation. The independently valid single-owner `Wait`/output-drain lifecycle behavior from current `main` is preserved.

This is intentionally a containment measure, not a claim that the regression risk is gone. Enabled Windows 10 commands still change from redirected pipes to a PTY. That can affect TTY detection, stdout/stderr merging, VT repaint behavior, and commands that expect a particular stdin model. The revised approach only narrows that risk to the Windows 10 environment where the bug has been reproduced and the fix has been validated; regressions remain possible for other commands and devices within Windows 10.

File: internal/runner/process.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	"io"
+	"os"
+	"strings"
+)
+
+type commandProcess interface {
+	Start() ([]io.ReadCloser, error)
+	Wait() error
+	PID() int
+	shouldSkipOutputLine(line string) bool
```
Comment (moisesrj97) [inline internal/runner/process.go]: The one I care most about. `shouldSkipOutputLine` is a conhost quirk, but putting it on `commandProcess` spreads it: `streamOutput` grew a `process` param (`runner.go:228`) and `execProcess` carries a `return false` that only satisfies the interface.

`conPTYOutput` already strips `\x1b[2J`/`\x1b[H` at `proc_windows.go:49`. Filtering there too deletes the interface method, the param and the Unix no-op. Slightly more work since that wrapper is byte-oriented and this filter is line-oriented, but nothing outside Windows should know this sequence exists.

File: internal/runner/process_exec.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+func (p *execProcess) PID() int {
+	return p.cmd.Process.Pid
+}
+
+func (p *execProcess) shouldSkipOutputLine(string) bool {
+	return false
+}
+
+func closeReaders(readers ...io.Closer) {
```
Comment (moisesrj97) [inline internal/runner/process_exec.go]: Same function as `closeProcessReaders` (`process.go:16`) under two names in one package. Keep the variadic `io.Closer` one and call `closeReaders(readers...)` in the drain path.

File: cmd/gomander/application_config.go [hunk 6]
```diff
@@ -0,0 +1,17 @@
+package main
+
+import "gomander/internal/runner"
+
+type applicationConfig struct {
+	configFolderName string
+	runner           runner.Config
+}
+
+var appConfig = applicationConfig{
```
Comment (moisesrj97) [inline cmd/gomander/application_config.go]: `buildDeps` is meant to be the single wiring point (`.claude/CLAUDE.md`, and main.go's own comment). A package global that `buildDeps` and `getDbFile` reach out to breaks that and removes the seam for a test-built runner config. Constructing it in `main` and passing it in costs the same.

File: cmd/gomander/application_config.go [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+	configFolderName: "gomander",
+	runner: runner.Config{
+		ConPTYEnvironments: []runner.HostEnvironment{
```
Comment (moisesrj97) [inline cmd/gomander/application_config.go]: Ship this empty so ConPTY is opt-in. It currently enables a path no test executes anywhere, and the follow-ups in the review body stop being merge blockers.

**Review by moisesrj97 — COMMENTED** [review]
Most of my comments are outside Windows, deliberately: Windows is a small share of our users, so the shared runner code carries more weight here than the ConPTY path.

One Windows change before merge: ship `ConPTYEnvironments` empty. CI never executes that path — `windows-latest` is Server 2022, which `classifyWindowsVersion` returns `""` for, so every behavioural test skips. Opt-in lets the refactor land now.

Follow-ups for when the flag goes on, not blocking:

- `proc_windows.go:116` — the child's stdin is the pseudoconsole pipe nobody writes to, and `InputPipe().Close()` only runs inside `Wait()`. Commands that read stdin block instead of getting EOF.
- `proc_windows.go:190` — a ConPTY client has no console window, so non-forced `taskkill` can't stop it; we always fall through to `/F /T` and the 5s grace is dead code. `\x03` to `InputPipe()` is the native equivalent.
- `proc_windows.go:182` — go-pty resolves a bare `cmd.exe` against `Dir`, i.e. the project directory. Use an absolute `%SystemRoot%` path.
- `proc_windows.go:136` — `Resize(32766, 1000)` makes conhost preallocate a large buffer per command; adds up across a group.
- `proc_windows.go:43` — `initialized = true` is set before `pending` is assigned, across a blocking `ReadBytes`.
- `proc_windows.go:81` — `RtlGetVersion` and the `CreatePseudoConsole` probe run per command start under the mutex, though both are constant. `sync.Once`.

File: internal/runner/process.go (hunk 4 above)
Comment (ailtariel-lin) [reply internal/runner/process.go]: Fixed in `950ec38`. ConPTY host-only line filtering now lives entirely in `conPTYOutput`; `shouldSkipOutputLine` was removed from `commandProcess`, along with the Unix no-op and the process parameter on `streamOutput`. Added a Windows adapter test covering initialization/title filtering while preserving command output.

File: internal/runner/process_exec.go (hunk 5 above)
Comment (ailtariel-lin) [reply internal/runner/process_exec.go]: Fixed in `950ec38`. Removed `closeProcessReaders` and use the single variadic `closeReaders` helper in the drain path. The helper is generic over `io.Closer` so an existing `[]io.ReadCloser` can be expanded directly without a slice conversion.

File: cmd/gomander/application_config.go (hunk 6 above)
Comment (ailtariel-lin) [reply cmd/gomander/application_config.go]: Fixed in `950ec38`. `applicationConfig` is now constructed in `main`, and the config-folder and runner values are passed explicitly into `configDB` and `buildDeps`; there is no package-global configuration instance.

File: cmd/gomander/application_config.go (hunk 7 above)
Comment (ailtariel-lin) [reply cmd/gomander/application_config.go]: Fixed in `950ec38`. `ConPTYEnvironments` now ships empty, so the default build retains the existing pipe runner and ConPTY requires an explicit application-configuration change. I also updated the PR title/body and changed `Closes #228` to `Related to #228` to reflect that default behavior.

**Comment by ailtariel-lin** [comment]
Addressed the merge-blocking review items in `950ec38`:

- `ConPTYEnvironments` now ships empty, so the default build retains the existing hidden `cmd /C` pipe path and ConPTY requires an explicit application-configuration change.
- `applicationConfig` is constructed in `main`; the config-folder and runner values are passed explicitly into `configDB` and `buildDeps` instead of being read from package-global state.
- ConPTY initialization and conhost title filtering now live entirely in `conPTYOutput`. The shared `commandProcess.shouldSkipOutputLine` method, Unix no-op, and extra `streamOutput` process parameter were removed.
- The duplicate reader-closing helpers were consolidated into one variadic helper.
- Added a Windows adapter regression test covering ConPTY initialization/title filtering while preserving command output.

I also updated the PR title and description to reflect that this is disabled-by-default, opt-in infrastructure, and changed `Closes #228` to `Related to #228`.

Validation completed:

- `go test ./internal/runner -count=1`
- `go vet ./internal/runner`
- `go test ./... -count=1`
- Linux amd64 runner test cross-compilation

The Windows-specific follow-ups listed in the review remain intentionally out of scope while the allowlist ships empty.

**Comment by moisesrj97** [comment]
One more concern, can this bug be reproduced in Windows 11?

**Comment by moisesrj97** [comment]
Closed due to inactivity
